File: legacy/hip3/src/hip3_oracle/aggregation.py
from __future__ import annotations

from decimal import Decimal
from typing import Iterable

from .config import FeedConfig
from .models import AggregatePrice, MarketStatus, Quote, RejectedQuote


BPS = Decimal("10000")


class AggregationError(RuntimeError):
    pass


def weighted_quantile(quotes: list[Quote], quantile: Decimal, *, key=lambda quote: quote.price) -> Decimal:
    if not quotes:
        raise AggregationError("cannot calculate a quantile without quotes")
    ordered = sorted(quotes, key=key)
    total = sum((quote.weight for quote in ordered), Decimal(0))
    target = total * quantile
    cumulative = Decimal(0)
    for quote in ordered:
        cumulative += quote.weight
        if cumulative >= target:
            return key(quote)
    return key(ordered[-1])


def _market_status(quotes: Iterable[Quote]) -> MarketStatus:
    statuses = {quote.market_status for quote in quotes}
    if MarketStatus.OPEN in statuses and MarketStatus.CLOSED in statuses:
        return MarketStatus.UNKNOWN
    if MarketStatus.CLOSED in statuses:
        return MarketStatus.CLOSED
    if MarketStatus.OPEN in statuses:
        return MarketStatus.OPEN
    return MarketStatus.UNKNOWN


class PriceAggregator:
    def aggregate(self, feed: FeedConfig, quotes: Iterable[Quote], *, now_ms: int) -> AggregatePrice:
        accepted: list[Quote] = []
        rejected: list[RejectedQuote] = []
        seen: set[str] = set()

        for quote in quotes:
            reason = self._validate_quote(feed, quote, now_ms, seen)
            if reason:
                rejected.append(RejectedQuote(quote.source, reason))
                continue
            seen.add(quote.source)
            accepted.append(quote)

        self._require_quorum(feed, accepted, "before outlier filtering")
        grouped = self._collapse_groups(accepted)
        center = weighted_quantile(grouped, Decimal("0.5"))
        deviations = [
            Quote(
                coin=quote.coin,
                source=quote.source,
                independence_group=quote.independence_group,
                price=abs(quote.price - center),
                observed_at_ms=quote.observed_at_ms,
                received_at_ms=quote.received_at_ms,
                weight=quote.weight,
            )
            for quote in grouped
        ]
        mad = weighted_quantile(deviations, Decimal("0.5"))
        minimum_band = center * feed.outlier_min_band_bps / BPS
        outlier_band = max(mad * feed.outlier_mad_multiplier, minimum_band)

        filtered_groups: list[Quote] = []
        rejected_groups: set[str] = set()
        for quote in grouped:
            if abs(quote.price - center) > outlier_band:
                rejected_groups.add(quote.independence_group)
            else:
                filtered_groups.append(quote)
        filtered = [quote for quote in accepted if quote.independence_group not in rejected_groups]
        rejected.extend(
            RejectedQuote(quote.source, "MAD outlier group")
            for quote in accepted
            if quote.independence_group in rejected_groups
        )
        self._require_quorum(feed, filtered, "after outlier filtering")

        price = weighted_quantile(filtered_groups, Decimal("0.5"))
        q25 = weighted_quantile(filtered_groups, Decimal("0.25"))
        q75 = weighted_quantile(filtered_groups, Decimal("0.75"))
        confidence = max(abs(price - q25), abs(q75 - price))
        confidence_bps = confidence * BPS / price
        if confidence_bps > feed.max_confidence_bps:
            raise AggregationError(
                f"{feed.coin} confidence {confidence_bps:.4f} bps exceeds {feed.max_confidence_bps} bps"
            )

        return AggregatePrice(
            coin=feed.coin,
            price=price,
            confidence=confidence,
            confidence_bps=confidence_bps,
            observed_at_ms=min(quote.observed_at_ms for quote in filtered),
            source_count=len(filtered),
            independent_group_count=len({quote.independence_group for quote in filtered}),
            market_status=_market_status(filtered),
            sources=tuple(sorted(quote.source for quote in filtered)),
            rejected=tuple(rejected),
        )

    @staticmethod
    def _collapse_groups(quotes: list[Quote]) -> list[Quote]:
        """Give one upstream group one vote even when it is exposed through multiple APIs."""
        by_group: dict[str, list[Quote]] = {}
        for quote in quotes:
            by_group.setdefault(quote.independence_group, []).append(quote)
        grouped: list[Quote] = []
        for group, members in by_group.items():
            grouped.append(
                Quote(
                    coin=members[0].coin,
                    source=f"group:{group}",
                    independence_group=group,
                    price=weighted_quantile(members, Decimal("0.5")),
                    observed_at_ms=min(member.observed_at_ms for member in members),
                    received_at_ms=max(member.received_at_ms for member in members),
                    # Multiple endpoints for one upstream must not accumulate voting weight.
                    weight=max(member.weight for member in members),
                    market_status=_market_status(members),
                )
            )
        return grouped

    @staticmethod
    def _validate_quote(feed: FeedConfig, quote: Quote, now_ms: int, seen: set[str]) -> str | None:
        if quote.coin != feed.coin:
            return "wrong coin"
        if quote.source not in feed.sources:
            return "source is not configured for feed"
        if quote.source in seen:
            return "duplicate source"
        if not quote.price.is_finite() or quote.price <= 0:
            return "price is not finite and positive"
        if not quote.weight.is_finite() or quote.weight <= 0:
            return "weight is not finite and positive"
        if quote.observed_at_ms > now_ms + feed.max_future_ms:
            return "future timestamp"
        if now_ms - quote.observed_at_ms > feed.max_source_age_ms:
            return "stale timestamp"
        if quote.bid is not None or quote.ask is not None:
            if quote.bid is None or quote.ask is None:
                return "bid and ask must be supplied together"
            if not quote.bid.is_finite() or not quote.ask.is_finite():
                return "invalid bid/ask"
            if quote.bid <= 0 or quote.ask <= 0 or quote.bid > quote.ask:
                return "invalid bid/ask"
            mid = (quote.bid + quote.ask) / Decimal(2)
            spread_bps = (quote.ask - quote.bid) * BPS / mid
            if spread_bps > feed.max_spread_bps:
                return "spread exceeds limit"
        return None

    @staticmethod
    def _require_quorum(feed: FeedConfig, quotes: list[Quote], stage: str) -> None:
        if len(quotes) < feed.min_sources:
            raise AggregationError(f"{feed.coin} has {len(quotes)} valid sources {stage}; requires {feed.min_sources}")
        groups = len({quote.independence_group for quote in quotes})
        if groups < feed.min_independent_groups:
            raise AggregationError(
                f"{feed.coin} has {groups} independent groups {stage}; requires {feed.min_independent_groups}"
            )

File: oracle-node/src/aggregation.rs
use crate::config::Config;
use anyhow::{ensure, Result};
use rust_decimal::{prelude::ToPrimitive, Decimal};
use serde::Serialize;
use std::{
    collections::{BTreeMap, HashSet},
    str::FromStr,
};

pub const SCALE: u128 = 100_000_000;
// Bound ensures intermediate arithmetic cannot overflow u128, even with weights and bps.
pub const MAX_PRICE: u128 = 1_000_000_000_000_000_000;
pub fn parse_price(s: &str) -> Result<u128> {
    let d = Decimal::from_str(s)?;
    ensure!(d > Decimal::ZERO, "price must be positive");
    let scaled = d
        .checked_mul(Decimal::from(SCALE))
        .ok_or_else(|| anyhow::anyhow!("price overflow"))?;
    ensure!(scaled.fract().is_zero(), "price exceeds 8 decimals");
    let p = scaled
        .to_u128()
        .ok_or_else(|| anyhow::anyhow!("invalid price"))?;
    ensure!(p <= MAX_PRICE, "price exceeds supported range");
    Ok(p)
}
pub fn display_price(p: u128) -> String {
    format!("{}.{:08}", p / SCALE, p % SCALE)
}
#[derive(Debug, Clone, Serialize)]
pub struct Quote {
    pub source: String,
    pub group: String,
    pub weight: u32,
    #[serde(serialize_with = "price_string")]
    pub price: u128,
    pub observed_at_ms: u64,
    pub received_at_ms: u64,
    pub latency_ms: u64,
    pub bid: Option<u128>,
    pub ask: Option<u128>,
}
fn price_string<S: serde::Serializer>(p: &u128, s: S) -> Result<S::Ok, S::Error> {
    s.serialize_str(&display_price(*p))
}
#[derive(Debug, Clone, Serialize)]
pub struct Aggregate {
    #[serde(serialize_with = "price_string")]
    pub price: u128,
    pub observed_at: u64,
    pub confidence_bps: u32,
    pub source_count: usize,
    pub group_count: usize,
    pub accepted: Vec<String>,
    pub rejected: Vec<String>,
}
fn quantile(values: &[(u128, u32)], numerator: u128, denominator: u128) -> u128 {
    let mut v = values.to_vec();
    v.sort_by_key(|x| x.0);
    let total: u128 = v.iter().map(|x| x.1 as u128).sum();
    let mut acc = 0;
    for (p, w) in v {
        acc += w as u128;
        if acc * denominator >= total * numerator {
            return p;
        }
    }
    unreachable!()
}
fn quorum(c: &Config, q: &[Quote]) -> Result<()> {
    ensure!(
        q.len() >= c.min_sources,
        "insufficient valid sources: {} < {}",
        q.len(),
        c.min_sources
    );
    let groups: HashSet<_> = q.iter().map(|q| &q.group).collect();
    ensure!(
        groups.len() >= c.min_groups,
        "insufficient independent groups: {} < {}",
        groups.len(),
        c.min_groups
    );
    Ok(())
}
pub fn aggregate(
    c: &Config,
    quotes: Vec<Quote>,
    now_ms: u64,
    previous: Option<u128>,
) -> Result<Aggregate> {
    let mut accepted = vec![];
    let mut rejected = vec![];
    let mut seen = HashSet::new();
    for q in quotes {
        let configured = c.sources.iter().find(|s| s.name == q.source);
        let valid_meta = configured
            .map(|s| s.group == q.group && s.weight == q.weight)
            .unwrap_or(false);
        let spread_ok = match (q.bid, q.ask) {
            (None, None) => true,
            (Some(b), Some(a)) if b > 0 && b <= a && a <= MAX_PRICE => {
                (a - b) * 20000 <= (a + b) * c.max_spread_bps as u128
            }
            _ => false,
        };
        if !valid_meta
            || !seen.insert(q.source.clone())
            || q.price == 0
            || q.price > MAX_PRICE
            || q.observed_at_ms > now_ms
            || now_ms.saturating_sub(q.observed_at_ms) > c.max_age_secs * 1000
            || !spread_ok
        {
            rejected.push(format!(
                "{}: invalid, duplicate, stale, future or excessive spread",
                q.source
            ));
        } else {
            accepted.push(q);
        }
    }
    quorum(c, &accepted)?;
    let mut grouped: BTreeMap<String, Vec<&Quote>> = BTreeMap::new();
    for q in &accepted {
        grouped.entry(q.group.clone()).or_default().push(q);
    }
    let groups: Vec<_> = grouped
        .iter()
        .map(|(name, members)| {
            let v: Vec<_> = members.iter().map(|q| (q.price, q.weight)).collect();
            (
                name.clone(),
                quantile(&v, 1, 2),
                members.iter().map(|q| q.weight).max().unwrap(),
            )
        })
        .collect();
    let v: Vec<_> = groups.iter().map(|(_, p, w)| (*p, *w)).collect();
    let center = quantile(&v, 1, 2);
    let deviations: Vec<_> = v.iter().map(|(p, w)| (p.abs_diff(center), *w)).collect();
    let band = (quantile(&deviations, 1, 2) * c.mad_multiplier as u128)
        .max(center * c.outlier_band_bps as u128 / 10000);
    let retained: HashSet<_> = groups
        .iter()
        .filter(|(_, p, _)| p.abs_diff(center) <= band)
        .map(|(g, _, _)| g.clone())
        .collect();
    accepted.retain(|q| {
        if retained.contains(&q.group) {
            true
        } else {
            rejected.push(format!("{}: MAD outlier group", q.source));
            false
        }
    });
    quorum(c, &accepted)?;
    let v: Vec<_> = groups
        .iter()
        .filter(|(g, _, _)| retained.contains(g))
        .map(|(_, p, w)| (*p, *w))
        .collect();
    let price = quantile(&v, 1, 2);
    let deviation = price
        .abs_diff(quantile(&v, 1, 4))
        .max(price.abs_diff(quantile(&v, 3, 4)));
    let bps = (deviation * 10000).div_ceil(price);
    ensure!(
        bps <= c.max_confidence_bps as u128,
        "confidence exceeds limit"
    );
    if let Some(p) = previous {
        ensure!(p > 0 && p <= MAX_PRICE, "invalid reference price");
        ensure!(
            price.abs_diff(p) * 10000 <= p * c.max_jump_bps as u128,
            "price jump circuit breaker"
        );
    }
    Ok(Aggregate {
        price,
        observed_at: accepted
            .iter()
            .map(|q| q.observed_at_ms / 1000)
            .min()
            .unwrap(),
        confidence_bps: bps as u32,
        source_count: accepted.len(),
        group_count: retained.len(),
        accepted: accepted.iter().map(|q| q.source.clone()).collect(),
        rejected,
    })
}
#[cfg(test)]
mod tests {
    use super::*;
    fn config() -> Config {
        serde_json::from_str(include_str!("../../config/demo.json")).unwrap()
    }
    fn quotes(c: &Config) -> Vec<Quote> {
        c.sources
            .iter()
            .map(|s| Quote {
                source: s.name.clone(),
                group: s.group.clone(),
                weight: s.weight,
                price: 3000 * SCALE,
                observed_at_ms: 100000,
                received_at_ms: 100000,
                latency_ms: 0,
                bid: None,
                ask: None,
            })
            .collect()
    }
    #[test]
    fn rejects_invalid_precision() {
        assert!(parse_price("NaN").is_err());
        assert!(parse_price("-1").is_err());
        assert!(parse_price("1.123456789").is_err());
        assert_eq!(parse_price("3000.25").unwrap(), 300025000000);
    }
    #[test]
    fn removes_outlier() {
        let c = config();
        let mut q = quotes(&c);
        q[3].price = 9000 * SCALE;
        let a = aggregate(&c, q, 100000, None).unwrap();
        assert_eq!(a.source_count, 3);
        assert_eq!(a.price, 3000 * SCALE);
    }
    #[test]
    fn stale_and_future_fail_quorum() {
        let c = config();
        let mut q = quotes(&c);
        q[0].observed_at_ms = 0;
        q[1].observed_at_ms = 100001;
        assert!(aggregate(&c, q, 100000, None).is_err());
    }
    #[test]
    fn same_group_not_independent() {
        let mut c = config();
        for s in &mut c.sources {
            s.group = "same".into();
        }
        assert!(aggregate(&c, quotes(&c), 100000, None).is_err());
    }
    #[test]
    fn jump_blocks() {
        let c = config();
        assert!(aggregate(&c, quotes(&c), 100000, Some(2000 * SCALE)).is_err());
    }
    #[test]
    fn duplicate_rejected() {
        let c = config();
        let q = quotes(&c);
        assert!(aggregate(&c, vec![q[0].clone(); 4], 100000, None).is_err());
    }
    #[test]
    fn confidence_blocks() {
        let mut c = config();
        c.max_confidence_bps = 1;
        let mut q = quotes(&c);
        q[2].price = 3001 * SCALE;
        q[3].price = 3002 * SCALE;
        assert!(aggregate(&c, q, 100000, None).is_err());
    }
    #[test]
    fn spread_blocks() {
        let c = config();
        let mut q = quotes(&c);
        for x in &mut q {
            x.bid = Some(2000 * SCALE);
            x.ask = Some(4000 * SCALE);
        }
        assert!(aggregate(&c, q, 100000, None).is_err());
    }
}
